src/core/metadata/qgslayermetadata.cpp

```diff
@@ -273,6 +278,312 @@ void QgsLayerMetadata::readFromLayer( const QgsMapLayer *layer )
+      oneAddress.country = addressElement.namedItem( QStringLiteral( "country" ) ).toElement().text();
+      oneAddress.postalCode = addressElement.namedItem( QStringLiteral( "postalCode" ) ).toElement().text();
+      oneAddress.type = addressElement.namedItem( QStringLiteral( "type" ) ).toElement().text();
+      addresses << oneAddress ;
+    }
+    oneContact.addresses = addresses;
+    addContact( oneContact );
+  }
+
+  // links
+  mnl = metadataElement.namedItem( QStringLiteral( "links" ) );
+  mne = mnl.toElement();
+  QDomNodeList el = mne.elementsByTagName( QStringLiteral( "link" ) );
+  for ( int i = 0; i < el.size(); i++ )
+  {
+    mne = el.at( i ).toElement();
+    QgsLayerMetadata::Link oneLink;
+    oneLink.name = mne.attribute( QStringLiteral( "name" ) );
+    oneLink.type = mne.attribute( QStringLiteral( "type" ) );
+    oneLink.url = mne.attribute( QStringLiteral( "url" ) );
+    oneLink.description = mne.attribute( QStringLiteral( "description" ) );
+    oneLink.format = mne.attribute( QStringLiteral( "format" ) );
+    oneLink.mimeType = mne.attribute( QStringLiteral( "mimeType" ) );
+    oneLink.size = mne.attribute( QStringLiteral( "size" ) );
+    addLink( oneLink );
+  }
+
+  return true;
+}
+
+bool QgsLayerMetadata::writeMetadataXml( QDomElement &metadataElement, QDomDocument &document ) const
+{
+  // identifier
+  QDomElement identifier = document.createElement( QStringLiteral( "identifier" ) );
+  QDomText identifierText = document.createTextNode( mIdentifier );
+  identifier.appendChild( identifierText );
+  metadataElement.appendChild( identifier );
+
+  // parent identifier
+  QDomElement parentIdentifier = document.createElement( QStringLiteral( "parentidentifier" ) );
+  QDomText parentIdentifierText = document.createTextNode( mParentIdentifier );
+  parentIdentifier.appendChild( parentIdentifierText );
+  metadataElement.appendChild( parentIdentifier );
+
+  // language
+  QDomElement language = document.createElement( QStringLiteral( "language" ) );
+  QDomText languageText = document.createTextNode( mLanguage );
+  language.appendChild( languageText );
+  metadataElement.appendChild( language );
+
+  // type
+  QDomElement type = document.createElement( QStringLiteral( "type" ) );
+  QDomText typeText = document.createTextNode( mType );
+  type.appendChild( typeText );
+  metadataElement.appendChild( type );
+
+  // title
+  QDomElement title = document.createElement( QStringLiteral( "title" ) );
+  QDomText titleText = document.createTextNode( mTitle );
+  title.appendChild( titleText );
+  metadataElement.appendChild( title );
+
+  // abstract
+  QDomElement abstract = document.createElement( QStringLiteral( "abstract" ) );
+  QDomText abstractText = document.createTextNode( mAbstract );
+  abstract.appendChild( abstractText );
+  metadataElement.appendChild( abstract );
+
+  // keywords
+  QMapIterator<QString, QStringList> i( keywords() );
```

Comment:
Not sure if this will cause a detach of mKeywords. It's probably safer here to change it to `i( mKeywords );` instead of calling the function to ensure that there's no copy involved.

Reply:
Sorry, but a quick question. Does `detach` mean a copy in memory? I tried to google it, but I could find only references to thread.

Reply:
It relates to Qt implicitly shared classes - see http://doc.qt.io/qt-5/implicit-sharing.html. In general copies of these classes are very fast, since the underlying data isn't copied. But when a detach is triggered (e.g. by calling a non-const method), then the (probably expensive) deep copy is incurred (a "detach"). You want to try to minimise detaches so that they only occur when there's no alternative.